### PR TITLE
Deprecate string + value overloads

### LINKED ISSUE #13281: string concatenation with ranges causes function promotion instead of coercion-to-string
Possible feature request. The following code uses string concatenation instead of variadic arguments (like `writeln`), but exhibits function promotion with ranges instead of a coercion of range to string. Is this intended or just an oversight?

```chapel
proc println(s: string) {
  writeln(s);
}

proc main() {
  writeln("test 1");
  writeln(1..10);
  
  writeln("\ntest 2");
  writeln("" + (1..10));
  
  writeln("\ntest 3");
  writeln("" + ((1..10) :string));
  
  writeln("\ntest 4");
  println("" + (1..10));
}
```
```terminal
test 1
1..10

test 2
1 2 3 4 5 6 7 8 9 10

test 3
1..10

test 4
6
7
8
9
10
1
2
3
4
5
```

I suspect if there was a coercion for ranges to strings, then the intended output of all `1..10` would occur, but I don't know how to test this hypothesis. I also don't know if there's unintended consequences with what is proposed.

### LINKED ISSUE #13409: What string + value concatenations should be supported by default, if any?
At present, Chapel supports mixed-type concatenations between strings and...

* ...numeric types
* ...enumerated types
* ...booleans

with the arguments in either order.  This can often be convenient (e.g., we particularly use them a lot in generating error messages), but it also feels a bit arbitrary (why just these types?  why not others?  or why any other than strings at all?)

This issue asks whether we should:

* not support (by default anyway) string concatenation with any types other than strings (users would have to write `"answer is: " + val:string` rather than `"answer is: " + val`)
* stick with the status quo
* support string concatenation with any type that supports being cast to string
* something else

For context, this issue came out of #13281 in which the status quo resulted in confusion.

### PR DESCRIPTION
**Pull request by bradcray**
This PR deprecates `+()` overloads taking string/value pairs on the argument that we are not consistent about when we support them and current thinking suggests that one of the arguments should be cast to a compatible type with the other rather than relying on these overloads.  This decision started from a conversation on issue #13281 and which then spawned issue #13409.

The PR adds deprecation warnings for the overloads and rewrites all existing tests to avoid the use of such operators, typically by adding explicit casts to string; or by using varargs in the common case of calls to writeln().

The PR also adds a new deprecation test to help us remember to remove the future (and ensure the warnings are working: test/deprecated/stringPlusVal.chpl.  And it contains a new future that I ran into when rewriting tests in which casting a string to a string on a remote locale results in an unclear error message: test/multilocale/strings/remoteStringCast.chpl.

Resolves #13281 
Resolves #13409

### CODE REVIEW

**Comment by bradcray** [comment]
The reviewer of this PR should be sure to review (I don't know that it's worthwhile or necessary to review every change to every test):
* modules/internal/String.chpl
* modules/standard/*
* modules/internal/*
* spec/*
* test/release/examples/*
* test/deprecated/stringPlusVal.*
* test/multilocale/strings/remoteStringCast.*

**Comment by bradcray** [comment]
Over on issue #13409, we've been discussing removing + overloads between strings and values, which this PR implements.  The following is a map from developers to code that they've been involved with in hopes that you will spot-check code you've been involved with (or any others you wish to) to ensure there are no surprises with the direction being taken here and the impact on existing code:

@daviditen:
* DateTime.chpl
* test/*/diten/*
* test/interop/fortran/*

@mppf:
* test/classes/delete-free/*
* test/*/ferguson/*
* sort-performance-explorer.chpl

@lydia-duncan @dlongnecke-cray :
* test/interop/C/*
* test/*/lydia/*

@vasslitvinov:
* test/library/packages/VectorList/*
* test/reductions/partial/utilities.chpl
* test/*/vass/*

@ronawho:
* test/parallel/taskCount/runningTaskCount/*
* test/runtime/configMatters/comm/unordered/*
* test/studies/bale/*

@benharsh:
* test/release/examples/benchmarks/isx/isx.chpl
* test/studies/isx/*.chpl

@cassella:
* test/io/cassella/readline.chpl

@oplambeck / @ben-albrecht :
* TOML.chpl
* tools/mason/MasonUtils.chpl

@krishnadey30 / @ben-albrecht:
* test/library/packages/UnitTest/*

**Comment by dmk42** [comment]
The only part I was worried about was the error message in the param version of `string.byte()`, but you've taken care of that.

File: spec/Expressions.tex [hunk 1]
```diff
@@ -1361,30 +1361,29 @@ \subsection{The String Concatenation Operator}
...
 \begin{chapeloutput}
-result: 3
+hi there
 \end{chapeloutput}
-where \chpl{i} is an integer appends the string representation of \chpl{i} to the
-string literal \chpl{"result: "}.  If \chpl{i} is \chpl{3}, then the resulting string
-would be \chpl{"result: 3"}.
-\begin{chapelnoprint}
-\end{chapelnoprint}
+will cause \chpl{z} to be a new string containing the value \chpl"hi
```
Comment (cassella) [inline spec/Expressions.tex]: Should that be `\chpl{"hi` at the end of the line?

**Comment by cassella** [comment]
`readline.chpl`: :+1:

**Comment by e-kayrakli** [comment]
(I thought I commented on this but apparently I didn't)

Note that `_bytes` type still supports `+`'ing with strings. I think even if we weren't taking this general direction, the `bytes` type shouldn't have supported that without explicitly decoding. Because we are slowly rolling in the `bytes` type and associated support, I can create a separate PR and keep this one for more established types.

**Comment by e-kayrakli** [comment]
Good point on coercion. `bytes+string` should be fine as long as the user is aware that the result is a `bytes` and not a `string`. My reflex is still trying to avoid supporting it (if possible) even in a world where strings can coerce to bytes. But I'll think more about it.

File: spec/Expressions.tex (hunk 1 above)
Comment (bradcray) [reply spec/Expressions.tex]: Sheesh, sloppy...  thanks for the catch!

**Comment by bradcray** [comment]
@LouisJenkinsCS: Yes, that should work as long as there's a cast from other.type to string (which there is for most/all built-in types).  And I believe it should take care of all the cases that are being deprecated here.

File: test/multilocale/strings/remoteStringCast.future [hunk 2]
```diff
@@ -0,0 +1,8 @@
+bug: string:string cast within on-clause gives follower error
+
+While this pattern isn't particularly interesting or useful as written
+here, it seems that it ought to be legal, yet it gives an error:
+"follower iterators accepting a non-POD argument by in-intent are not
+implemented".  The pattern is more interesting/useful in the context
+of a generic function where you want to cast an incoming argument to
+'x' regardless of its type and want it to also work with strings.
```
Comment (vasslitvinov) [inline test/multilocale/strings/remoteStringCast.future]: Add a proper newline at the end?

Aside: I may be a good person to take a look at it, as I was the one who added this error to the compiler.

File: modules/internal/String.chpl [hunk 3]
```diff
@@ -2215,14 +2215,20 @@ module String {
     return ret;
   }
 
+  private proc stringValDeprecated() {
+    compilerWarning("'+' is deprecated for string/value pairs; please cast the value to string instead");
+  }
```
Comment (vasslitvinov) [inline modules/internal/String.chpl]: While this wording is reasonable, please consider changing it slightly to something like this: `+ is deprecated for string/non-string pairs; please add an explicit cast from the non-string value to string`.

File: modules/internal/ChapelTuple.chpl [hunk 4]
```diff
@@ -100,7 +100,7 @@ module ChapelTuple {
   pragma "no doc"
   proc *(param p: uint, type t) type {
     if p > max(int) then
-      compilerError("Tuples of size >" + max(int) + " are not currently supported");
+      compilerError("Tuples of size >" + max(int):string + " are not currently supported");
```
Comment (vasslitvinov) [inline modules/internal/ChapelTuple.chpl]: Aside: we should probably lower this limit way down from max(int). We generate one field per each tuple component, so I have seen the compiler slow down significantly at 200k.

File: modules/internal/String.chpl (hunk 3 above)
Comment (bradcray) [reply modules/internal/String.chpl]: How about `'+' between strings and non-strings is deprecated; consider explicitly casting the non-string argument to a string`?

File: modules/internal/ChapelTuple.chpl (hunk 4 above)
Comment (bradcray) [reply modules/internal/ChapelTuple.chpl]: I'm surprised we could get to 200k.  I thought it slowed down pretty badly even after double digits.

File: modules/internal/String.chpl (hunk 3 above)
Comment (vasslitvinov) [reply modules/internal/String.chpl]: Works for me!

File: modules/internal/ChapelTuple.chpl (hunk 4 above)
Comment (vasslitvinov) [reply modules/internal/ChapelTuple.chpl]: It never passed resolution at 200k. So, putting a stop after double digits would be fine by me as well.

**Comment by bradcray** [comment]
Anyone who was tagged to review tests that didn't get a chance, feel free to weigh in post-merge if you have any concerns.

**Pull request merged**